server/routerlicious/packages/services-utils/src/throttlerMiddleware.ts

```diff
@@ -0,0 +1,101 @@
+    /**
+     * Relative weight of request amongst other requests with same suffix.
+     */
+    weight: number;
+
+    /**
+     * Distinguishes throttle id amongst other tracked ids with same suffix.
+     * For example, this could be a tenantId, clientId, or endpoint name.
+     *
+     * Can be a function that takes in the `express.Request` and returns the prefix as a string,
+     * which is useful for getting the id prefix from route params, such as `tenantId`.
+     */
+    throttleIdPrefix?: string | ((req: Request) => string);
+
+    /**
+     * Distinguishes throttle id amongst other tracked ids with same prefix.
+     * For example, this could be "HistorianRest", "AlfredRest", "OpenSocketConn", or "SubmitOp".
+     */
+    throttleIdSuffix?: string;
+
+    /**
+     * If true, will decrement tracked throttler count for the given id when the response has been sent.
+     */
+    decrementOnFinish: boolean;
+}
+
+const defaultThrottleMiddlewareOptions: IThrottleMiddlewareOptions = {
+    weight: 1,
+    throttleIdPrefix: undefined,
+    throttleIdSuffix: undefined,
+    decrementOnFinish: false,
+};
+
+const getThrottleId = (req: Request, throttleOptions: IThrottleMiddlewareOptions) => {
+    let prefix: string | undefined;
+    if (typeof throttleOptions.throttleIdPrefix === "function") {
+        prefix = throttleOptions.throttleIdPrefix(req);
+    } else {
+        prefix = throttleOptions.throttleIdPrefix;
+    }
+
+    if (prefix && throttleOptions.throttleIdSuffix) {
+        return `${prefix}_${throttleOptions.throttleIdSuffix}`;
+    }
+    return prefix || throttleOptions.throttleIdSuffix || "-";
+};
+
+/**
+ * Express middleware for API throttling.
+ */
+export function throttle(
+    throttler: IThrottler,
+    logger?: ILogger,
+    options?: Partial<IThrottleMiddlewareOptions>): RequestHandler {
+        const throttleOptions = {
+            ...defaultThrottleMiddlewareOptions,
+            ...options,
+        };
+
+        return (req, res, next) => {
+            const throttleId = getThrottleId(req, throttleOptions);
+            const messageMetaData = {
+                key: throttleId,
+                weight: throttleOptions.weight,
+                event_type: "throttling",
+            };
+
+            logger?.info(`Incrementing throttle count: ${throttleId}`, { messageMetaData });
+            try {
+                throttler.incrementCount(throttleId, throttleOptions.weight);
+            } catch (e) {
+                if (e instanceof ThrottlingError) {
+                    logger?.info(`Throttled: ${throttleId}`, { messageMetaData: {
+                        ...messageMetaData,
+                        reason: e.message,
+                        retryAfterInSeconds: e.retryAfter,
+                    } });
+                    return res.status(e.code).json(e);
+                }
+            }
+
+            if (throttleOptions.decrementOnFinish) {
+                onFinished(res, () => {
```

Comment:
Why are we decrementing the count here?
It should be decremented when the caller of throtterMiddleware is done processing the request.

Reply:
That's what `onFinished` does. When the request is done processing and the response is sent, this callback is fired

Reply:
Got it. But this for Http requests only? What about socket.io processing count decrement?

Reply:
I think this middleware is only for http requests?

Reply:
Correct, this middleware is for HTTP only. Talked with @sharmakhushboo offline, and we agreed on removing `decrementCount` functionality from HTTP request throttling. It will only be used for open/closed socket connections